### PR TITLE
support update of job queue

### PR DESCRIPTION
**Pull request by grondo**
This is an experimental WIP that allows users to move a pending job to a new queue.

The list of valid queues and their current status (enabled/disabled) is currently private to `job-manager/queue.c`, so it isn't straightforward to add an `update-queue` plugin. The plugin would have to send an RPC back to the job-manager itself to query the list of queues and their states (and there's currently no way to do asynchronous work in one of these plugins), or a `jobtap` API call to "unpack" the queues or otherwise query them would be required.

Instead, this PR proposes a way for job manager components to register builtins directly from within their initialization routines, and pass some opaque data along that can be passed to all jobtap callbacks. This seems a little backwards, since it exposes internals to jobtap plugins, which slightly breaks the design, but it happens to be very useful for the specific case of allowing updates for an attribute of jobspec which is managed by a private job manager component.

This allows the `.update-queue` plugin to easily access queue-specific data without RPCs or ad-hoc jobtap functions that expose some internals of the queue module. It also keeps all the queue related code in one place, and reduces the amount of boiler plate code to a minimum, while still allowing the "plugin" to be unloaded to prevent queue updates, or reloaded to re-enable them. If other queue restrictions are added in the future (e.g. access controls) these can easily be supported in the `.update-queue` plugin since it is internal to `job-manager/queue.c`.

Of course the drawback is that someone could never write a truly external queue-update plugin since we've not exposed those internals for general jobtap use. Since that is a pretty big caveat, I'm just proposing this PR as an experimental WIP at this time to get feedback on this particular design choice.

Either way, the next step here is to add a way for the plugin to request that a feasibility check be issued after the update has been allowed and before it is applied. This will require changes to `job-manager/update.c` itself, since the feasibility check requires an RPC to the scheduler.

Example:
```console
$ jobid=$(flux submit --urgency=hold -q batch hostname)
$ flux jobs -no {queue} $jobid
batch
$ flux update $jobid queue=debug
$ flux jobs -no {queue} $jobid
debug
$ flux update $jobid queue=foo
flux-update: ERROR: 'foo' is not a valid queue
$ flux queue disable -q batch testing update of disabled queue
batch: Job submission is disabled: testing update of disabled queue
$ flux update $jobid queue=batch
flux-update: ERROR: job submission to batch is disabled: testing update of disabled queue
```

### CODE REVIEW

**Comment by garlick** [comment]
> register builtins directly from within their initialization routines, and pass some opaque data along that can be passed to all jobtap callbacks

Oh that seems like a great idea!  If it makes this type of extension easier to use and/or maintain compared to a non-plugin implementation, then the capability sounds pretty useful.

> the drawback is that someone could never write a truly external queue-update plugin since we've not exposed those internals for general jobtap use

That doesn't seem like a biggie to me?  (Or am I missing some possible use case?)

Just beginning to look at this, but should there be a call to `flux_jobtap_jobspec_update_pack()` in the `.update-queue` plugin?

**Comment by grondo** [comment]
> Just beginning to look at this, but should there be a call to flux_jobtap_jobspec_update_pack() in the .update-queue plugin?

No. The way these `job.update.*` callbacks work is that the callback checks if the update is _allowable_. If it is, the callback should return `0`, i.e. success, if not `-1` is returned with an error message set via `flux_jobtap_error()`.

If there is no `job.update.` callback for the requested update, then the update is rejected. So updates have to be specifically allowed.

Once an update is allowed, then the update service in the job manager calls the full `job.validate` plugin stack to ensure the resulting updates are valid in addition to allowed.

**Comment by grondo** [comment]
> That doesn't seem like a biggie to me? (Or am I missing some possible use case?)

I can't think of one, but wanted to bring it up in case someone else had an objection or could think of a case where it would be better to somehow expose the internal queue states via a `flux_jobtap_*` call or set of calls.

File: src/modules/job-manager/jobtap-internal.h
Comment (chu11) [inline src/modules/job-manager/jobtap-internal.h:106]: commit message typo "opqaque"

**Comment by chu11** [comment]
My 10 second dumb thought, is was why can't current queues and their statuses be passed to jobtap plugins as args?

like could there be a new function `json_t *queue_status(struct queue *queue)` that returns a json dictionary of `queuename:status` pairs, and that gets passed to the jobtap plugins as args?

**Comment by grondo** [comment]
> My 10 second dumb thought, is was why can't current queues and their statuses be passed to jobtap plugins as args?

I had something similar in an initial implementation. Problems as I saw them with this approach is:

 1. It adds a lot more packing work to every jobap callback for a use case currently needed by only one callback that will be rarely used
 2. It adds more boilerplate to the queue update plugin.
 3. As more queue specific internal data is added to `struct queue` and `struct jobq` this would additionally need to be exposed packed in every jobtap callback
 4. Other job manager internals may want to allow updates or register other plugins with similar access to internal structures. This adds a way to do that in general, whereas passing via jobtap plugin args would require adding those packing functions to each location. (Not impossible, just more work and over time the plugin args could get very large)

Of course the benefit of your suggestion is that it would work for externally managed plugins. :shrug:

**Comment by chu11** [comment]
> It adds a lot more packing work to every jobap callback for a use case currently needed by only one callback that will be rarely used

True.  I figured we could cache the statuses b/c queue changes should be rare, leading to minimal extra packing.  But minimal != zero :P

**Comment by grondo** [comment]
Jobtap plugin arguments are constructed for each jobtap plugin callback.

My initial implementation had something like a `flux_jobtap_queue_info_unpack(3)` interface function, but that required a fair bit of work, packing and unpacking on the job manager and plugin side, and I didn't relish the idea that `jobtap.h` might have a large growth of these functions over time that have to reach into otherwise encapsulated interfaces of the job manager.

**Comment by grondo** [comment]
Ok, this WIP PR is now more fully fleshed out with a proposed full implementation.

Changes include:
 - The job update service now allows plugins to request a feasibility check for an update by setting a `feasibility` flag in the plugin out arguments. This not only allows the asynchronous feasibility check to occur, but also ensures that feasibility is tested against the update in its entirety. Note that if the scheduler isn't loaded the feasibility rpc will not succeed and the update will be denied.
 - The update-queue plugin requests a feasibility check of the updated jobspec.
 - Plugins can set additional required updates by adding an `updates` object to the plugin out arguments.
 - The update-queue plugin checks current job constraints, and if the constraints don't exactly match current queue constraints it rejects the update. If the constraints match, then the update-queue plugin overwrites current constraints with new queue constraints via the `updates` plugin output argument.

**Comment by grondo** [comment]
Note: before this can be merged, the flux-accounting priority plugin will have to be updated to support queue updates.

**Comment by grondo** [comment]
Note: This is pretty much done, just have to sync up with flux-accounting. However, removing WIP so we can get the implementation reviewd.

File: doc/man7/flux-jobtap-plugins.rst
Comment (chu11) [inline doc/man7/flux-jobtap-plugins.rst:275]: thee -> the

File: t/t2291-job-update-queue.t
Comment (chu11) [inline t/t2291-job-update-queue.t:87]: haven't tried yet, but could test `flux job list` to see if `job-list` got update (here and perhaps few other places)?  I think it should work ... if it doesn't, can be a TODO item for me to minimally add some tests in t2260-job-list.

Note, adding here could be annoying as you gotta spin a bit to wait for job-list eventual consistency.

**Review by chu11 — COMMENTED** [review]
first pass, nothing big i noticed, mostly smaller stuff

File: t/t2291-job-update-queue.t
Comment (grondo) [reply t/t2291-job-update-queue.t:87]: Yeah, I didn't do it because of the raciness, and job update processing in job-list is tested elsewhere so it could be considered redundant here.

File: src/modules/job-manager/update.c
Comment (grondo) [reply src/modules/job-manager/update.c:219]: `pending_request_create()` calls `update_request_create()`, or perhaps I'm misunderstanding the suggestion?

File: src/modules/job-manager/queue.c
Comment (grondo) [reply src/modules/job-manager/queue.c:853]: I started pulling this out into its own function for more clarity, then realized that there are some bugs here when a queue does not have a configured constraint. This will require a bit more code and tests, so working on that now.

**Comment by grondo** [comment]
I've addressed comments so far and rebased on current master.

Additionally, fixed handling of queues without constraints and added tests with constraint-less queue.

**Comment by grondo** [comment]
Ok, @cmoussa1 has proposed flux-framework/flux-accounting#389 which implements the update logic for queues (and thus tests depend on this PR getting merged).

File: src/modules/job-manager/update.c
Comment (garlick) [inline src/modules/job-manager/update.c:85]: An inline comment that `handle` is a zlistx handle for this item in `update->pending_requests` might be helpful.

File: src/modules/job-manager/update.c
Comment (garlick) [inline src/modules/job-manager/update.c:258]: Would the future be leaked if job-manager were unloaded with a feasibility check in progress?

File: src/modules/job-manager/update.c
Comment (garlick) [inline src/modules/job-manager/update.c:413]: A better log message might be "error responding to job-manager.update request".

File: src/modules/job-manager/queue.c
Comment (garlick) [inline src/modules/job-manager/queue.c:49]: Commit title:   totally minor but I think we've been using "topic/subtopic:" where a subtopic is needed.  "job-manager: queue:" doesn't  fit that pattern.

File: src/modules/job-manager/queue.c
Comment (garlick) [inline src/modules/job-manager/queue.c:85]: Suggestion: rename parameter `value` to `config` to indicate that it's the queue configuration for the named queue.

File: src/modules/job-manager/queue.c
Comment (garlick) [inline src/modules/job-manager/queue.c:27]: Excellent commit message!

File: src/modules/job-manager/queue.c
Comment (garlick) [inline src/modules/job-manager/queue.c:792]: I don't think you have to compare the two object sizes since [json_equal()](https://jansson.readthedocs.io/en/latest/apiref.html#c.json_equal) already requires that they have the same keys, per the jansson  [equality](https://jansson.readthedocs.io/en/latest/apiref.html#equality) definition.

File: t/t2291-job-update-queue.t
Comment (chu11) [inline t/t2291-job-update-queue.t]: should we add a variant test where there are no requires/constraints within a queue?

Comment (grondo) [reply t/t2291-job-update-queue.t]: Oh, yeah. I swear I did that, I wonder if I just didn't push the result. Let me look around...

Comment (grondo) [reply t/t2291-job-update-queue.t]: Oops, looks like it got squashed into the wrong commit :facepalm: I'll fix that.

File: t/t2291-job-update-queue.t
Comment (garlick) [inline t/t2291-job-update-queue.t]: Looks like maybe the test changes should be in a different commit?

File: t/t2291-job-update-queue.t
Comment (grondo) [reply t/t2291-job-update-queue.t]: Ok, moved the changes to the right commit and force-pushed the result.

**Review by garlick — APPROVED** [review]
Nice work!

I had a few comments but I leave it to you decide what is worth addressing before this gets merged.

File: t/t2291-job-update-queue.t
Comment (chu11) [inline t/t2291-job-update-queue.t:154]: typo, double apostraphe

File: t/t2291-job-update-queue.t
Comment (chu11) [reply t/t2291-job-update-queue.t]: changes LGTM!

File: src/modules/job-manager/queue.c
Comment (garlick) [reply src/modules/job-manager/queue.c:792]: Oh I just realized I misread that line of code and you're checking for empty.  Well I would assume two empty objects would be considered equal by `json_equal()` but maybe you've actually found that not to be the case!

File: src/modules/job-manager/queue.c
Comment (grondo) [reply src/modules/job-manager/queue.c:49]: Hm, I think this seems to be personal preference:

```
 grondo@pi3:~/git/flux-core.git$ git log | grep -c job-manager:.*: 
47
 grondo@pi3:~/git/flux-core.git$ git log | grep -c job-manager/ 
57
```

BTW, I don't necessarily like `topic/subtopic` because when you search for `topic: ` thinking you'll find all matches for `topic` you are mistaken :-)

File: src/modules/job-manager/queue.c
Comment (grondo) [reply src/modules/job-manager/queue.c:792]: No, the issue is that `json_object_size()` handles `NULL` (returns 0) and `json_equal()` does not (IIRC). We could check `c1 == NULL && c2 == NULL` but then empty objects would be unnecessarily passed to `json_equal()` for comparison, it seemed better to do that in one go... However, if it is confusing maybe the former is better code!

Comment (grondo) [reply src/modules/job-manager/queue.c:792]: Just verified if either argument is `NULL` then `json_equal()` will return false, however here we want `NULL == NULL` to return true. Maybe just explicitly testing for `NULL == NULL` would be least confusing...

File: src/modules/job-manager/update.c
Comment (grondo) [reply src/modules/job-manager/update.c:258]: Ah, yeah. The whole reason for the pending requests list was to do this cleanup, then I forgot to add the future there! :facepalm:

File: src/modules/job-manager/update.c
Comment (grondo) [reply src/modules/job-manager/update.c:413]: Sounds good. However, this is basically copying existing idiom in the job-manager, shall we clean up everywhere?

```
 grondo@pi3:~/git/flux-core.git/src/modules/job-manager$ git grep "%s: flux_respond_error"
alloc.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
alloc.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
alloc.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
annotate.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
drain.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
drain.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
drain.c:                flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
getattr.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
job-manager.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
job-manager.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
kill.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
kill.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
list.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
raise.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
raise.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
start.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
submit.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
update.c:        flux_log_error (ctx->h, "%s: flux_respond_error", __FUNCTION__);
update.c:                            "%s: flux_respond_error",
update.c:        flux_log_error (ctx->h, "%s: flux_respond_error", __FUNCTION__);
update.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
update.c:                            "%s: flux_respond_error",
urgency.c:        flux_log_error (h, "%s: flux_respond_error", __FUNCTION__);
wait.c:                flux_log_error (h, "%s: flux_respond_error", __func__);
wait.c:        flux_log_error (h, "%s: flux_respond_error", __func__);
```

Comment (grondo) [reply src/modules/job-manager/update.c:413]: Although now I see you were pointing out this instance as a special case. Seems reasonable.

File: t/t2291-job-update-queue.t
Comment (grondo) [reply t/t2291-job-update-queue.t:154]: oof!

**Comment by grondo** [comment]
Ok, I think I've addressed the various comments

File: src/modules/job-manager/update.c
Comment (garlick) [reply src/modules/job-manager/update.c]: Relying on __FUNCTION__ to provide the context for the error is not always helpful, so I've been shying away from that style of late, in new code at least.  No big deal though.

Comment (grondo) [reply src/modules/job-manager/update.c]: I completely agree. It is especially maddening when you try to search for an error from the logs but get no hits because of the use of `__FUNCTION__`.

It did seem reasonable to fix that one spot, so I did so in the latest commit. The other uses of `__FUNCTION__` are pervasive (almost every return from a request handler uses this idiom in the codebase) so I didn't touch those.

**Comment by grondo** [comment]
Ok to set MWP or @chu11 did you want another pass?

**Comment by chu11** [comment]
Go for it!

**Pull request merged**
